diff --git a/py/dynruntime.mk b/py/dynruntime.mk
index db06d41e7..a75e72c0e 100644
--- a/py/dynruntime.mk
+++ b/py/dynruntime.mk
@@ -133,7 +133,7 @@ $(BUILD)/%.mpy: %.py
 	$(Q)$(MPY_CROSS) $(MPY_CROSS_FLAGS) -o $@ $<
 
 # Build native .mpy from object files
-$(BUILD)/$(MOD).native.mpy: $(SRC_O)
+$(BUILD)/$(MOD).native.mpy: $(SRC_O) $(LINK_EXTRA)
 	$(ECHO) "LINK $<"
 	$(Q)$(MPY_LD) --arch $(ARCH) --qstrs $(CONFIG_H) -o $@ $^
 
diff --git a/tools/mpy_ld.py b/tools/mpy_ld.py
index 0dac69e20..7bda0d694 100755
--- a/tools/mpy_ld.py
+++ b/tools/mpy_ld.py
@@ -75,6 +75,8 @@ R_ARM_THM_JUMP24 = 30
 R_X86_64_GOTPCREL = 9
 R_X86_64_REX_GOTPCRELX = 42
 R_386_GOT32X = 43
+R_XTENSA_PDIFF32 = 59
+R_XTENSA_NDIFF32 = 62
 
 ################################################################################
 # Architecture configuration
@@ -564,7 +566,7 @@ def do_relocation_text(env, text_addr, r):
         reloc = addr - r_offset
         reloc_type = "xtensa_l32r"
 
-    elif env.arch.name == "EM_XTENSA" and r_info_type == R_XTENSA_DIFF32:
+    elif env.arch.name == "EM_XTENSA" and r_info_type in (R_XTENSA_DIFF32, R_XTENSA_PDIFF32, R_XTENSA_NDIFF32):
         if s.section.name.startswith(".text"):
             # it looks like R_XTENSA_DIFF32 into .text is already correctly relocated
             return
@@ -627,6 +629,8 @@ def do_relocation_data(env, text_addr, r):
         or env.arch.name == "EM_XTENSA"
         and r_info_type == R_XTENSA_32
     ):
+        if hasattr(s, "resolved"):
+            s = s.resolved
         # Relocation in data.rel.ro to internal/external symbol
         if env.arch.word_size == 4:
             struct_type = "<I"
@@ -957,7 +961,7 @@ def build_mpy(env, entry_offset, fmpy, native_qstr_vals, native_qstr_objs):
     for base, addr, kind in env.mpy_relocs:
         if isinstance(kind, str) and kind.startswith(".text"):
             kind = 0
-        elif kind in (".rodata", ".data.rel.ro"):
+        elif isinstance(kind, str) and (kind.startswith(".rodata") or kind.startswith(".data.rel.ro")):
             if env.arch.separate_rodata:
                 kind = rodata_const_table_idx
             else:
