diff --git a/py/dynruntime.mk b/py/dynruntime.mk
index db06d41e7..a75e72c0e 100644
--- a/py/dynruntime.mk
+++ b/py/dynruntime.mk
@@ -133,7 +133,7 @@ $(BUILD)/%.mpy: %.py
 	$(Q)$(MPY_CROSS) $(MPY_CROSS_FLAGS) -o $@ $<
 
 # Build native .mpy from object files
-$(BUILD)/$(MOD).native.mpy: $(SRC_O)
+$(BUILD)/$(MOD).native.mpy: $(SRC_O) $(LINK_EXTRA)
 	$(ECHO) "LINK $<"
 	$(Q)$(MPY_LD) --arch $(ARCH) --qstrs $(CONFIG_H) -o $@ $^
 
